### PR TITLE
Image dataset

### PR DESCRIPTION
**Pull request by markusnagel**
Added a new dataset which directly reads images from file. This is useful for data which is not frequently/recurrently used since it saves the need to convert it before to h5py. As input it expects a regular expression which gives great flexibility on how the images can be arranged on disk.

### CODE REVIEW

File: fuel/datasets/image.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    Parameters
+    ----------
+    pattern : basestring
+        The pattern which describes the path to all images to use. This is
+        expected in form of a regular excursion. For example
+        '/this/path/*jpeg' will use all jpeg images in the folder '/this/path'.
+        The images will be given in source ``images`` and it will also
+        give the source ``fiel_paths`` which contains the exact path of
+        the loaded image.
+    load_in_memory: bool, default=True
```
Comment (aukejw) [inline fuel/datasets/image.py]: don't forget to surround the ':' with spaces

Comment (aukejw) [inline fuel/datasets/image.py]: don't forget to surround the ':' with spaces!

File: fuel/datasets/image.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    """Creates a dataset from raw images on disk.
+
+    Parameters
+    ----------
+    pattern : basestring
+        The pattern which describes the path to all images to use. This is
+        expected in form of a regular excursion. For example
+        '/this/path/*jpeg' will use all jpeg images in the folder '/this/path'.
+        The images will be given in source ``images`` and it will also
+        give the source ``fiel_paths`` which contains the exact path of
```
Comment (aukejw) [inline fuel/datasets/image.py]: Typo `fiel_paths`

File: fuel/datasets/image.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+def _load_images(file_paths):
+    images = []
+    for fp in file_paths:
+        img = Image.open(fp)
+        images.append(numpy.array(img))
+    return images
+
+
+def _list_fancy_indexing(iterable, request):
```
Comment (aukejw) [inline fuel/datasets/image.py]: Can you add a docstring? Its purpose is not immediately clear from the name of the method.

**Review by aukejw — CHANGES_REQUESTED** [review]
Tests are missing, other than that some minor comments.

**Review by aukejw — APPROVED** [review]

File: tests/test_image_dataset.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        assert ds.num_examples == 3
+        ds = ImagesFromFile('{}/*.jpeg'.format(self.ds_path), start=2, stop=3)
+        assert ds.num_examples == 1
+
+    def test_get_data_dynamic(self):
+        ds = ImagesFromFile('{}/*.jpeg'.format(self.ds_path),
+                            load_in_memory=False)
+        stream = DataStream(ds, iteration_scheme=ShuffledScheme(
+            ds.num_examples, batch_size=10))
+        for imgs, _ in stream.get_epoch_iterator():
```
Comment (aukejw) [inline tests/test_image_dataset.py]: Final remark: can you assert that the contents of the image_path source match the dummy image paths? We now know that the batch and its content are of the correct shape, but not that it actually contains the images we were looking for..

File: tests/test_image_dataset.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+            assert len(imgs) == 5
+            assert imgs[0].shape == (512, 512, 3)
+
+    def test_get_data_inmemory(self):
+        ds = ImagesFromFile('{}/*.jpeg'.format(self.ds_path),
+                            load_in_memory=True)
+        stream = DataStream(ds, iteration_scheme=ShuffledScheme(
+            ds.num_examples, batch_size=10))
+        for imgs, _ in stream.get_epoch_iterator():
```
Comment (aukejw) [inline tests/test_image_dataset.py]: And here as well, of course.

**Review by aukejw — CHANGES_REQUESTED** [review]
Didn't notice it before, but the tests do not actually check whether the right thing is returned.

File: tests/test_image_dataset.py (hunk 4 above)
Comment (markusnagel) [reply tests/test_image_dataset.py]: Well having the correct number of images and the correct shape is a very strong indicator that we loaded the correct dataset. But I can check the 'image_paths' for the reason that that gets correctly set.

Comment (aukejw) [reply tests/test_image_dataset.py]: Please do, then we're certain that it's not the exact same image path being returned X times.

File: tests/test_image_dataset.py [hunk 6]
```diff
@@ -34,15 +34,17 @@ def test_get_data_dynamic(self):
                             load_in_memory=False)
         stream = DataStream(ds, iteration_scheme=ShuffledScheme(
             ds.num_examples, batch_size=10))
-        for imgs, _ in stream.get_epoch_iterator():
+        for imgs, file_paths in stream.get_epoch_iterator():
             assert len(imgs) == 5
             assert imgs[0].shape == (512, 512, 3)
+            assert self.ds_path in file_paths[0]
```
Comment (aukejw) [inline tests/test_image_dataset.py]: Why not `assert all(file_path in dummy_image_paths for image_path in file_paths)`? I understand that you don't want to test glob functionality, but given that there is a custom function `_list_fancy_indexing` in between it wouldn't hurt to actually check the correctness of the output.

**Review by aukejw — APPROVED** [review]

**Pull request merged**